File: app/src/main/java/com/shuyu/apprecycler/chat/detail/view/ImageBubbleLayout.java
package com.shuyu.apprecycler.chat.detail.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.RectF;
import android.graphics.Region;
import android.util.AttributeSet;
import android.view.View;

import com.daasuu.bl.ArrowDirection;
import com.daasuu.bl.BubbleLayout;

/**
 * 转为图片转换的布局
 */
public class ImageBubbleLayout extends BubbleLayout {

    private RectF mRect;
    private Path mPath = new Path();
    private Paint mPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
    private Path mStrokePath;
    private Paint mStrokePaint;

    public ImageBubbleLayout(Context context) {
        this(context, null, 0);
    }

    public ImageBubbleLayout(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public ImageBubbleLayout(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        mPaint.setColor(getBubbleColor());
    }


    @Override
    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
        super.onLayout(changed, left, top, right, bottom);
        int paddingLeft = getPaddingLeft();
        int paddingRight = getPaddingRight();
        int paddingTop = getPaddingTop();
        int paddingBottom = getPaddingBottom();
        switch (getArrowDirection()) {
            case LEFT:
                paddingLeft -= getArrowWidth();
                break;
            case RIGHT:
                paddingRight -= getArrowWidth();
                break;
            case TOP:
                paddingTop -= getArrowHeight();
                break;
            case BOTTOM:
                paddingBottom -= getArrowHeight();
                break;
        }
        setPadding(paddingLeft, paddingTop, paddingRight, paddingBottom);
        initDrawable(0, getWidth(), 0, getHeight());
    }


    @Override
    protected boolean drawChild(Canvas canvas, View child, long drawingTime) {
        drawPath(canvas);
        return super.drawChild(canvas, child, drawingTime);
    }

    private void drawPath(Canvas canvas) {
        if (getStrokeWidth() > 0) {
            canvas.drawPath(mStrokePath, mStrokePaint);
        }
        canvas.clipPath(mPath, Region.Op.INTERSECT);
    }


    private void initDrawable(int left, int right, int top, int bottom) {
        if (right < left || bottom < top) return;
        mRect = new RectF(left, top, right, bottom);
        if (getStrokeWidth() > 0) {
            mStrokePaint = new Paint(Paint.ANTI_ALIAS_FLAG);
            mStrokePaint.setColor(getStrokeColor());
            mStrokePath = new Path();

            initPath(getArrowDirection(), mPath, getStrokeWidth());
            initPath(getArrowDirection(), mStrokePath, 0);
        } else {
            initPath(getArrowDirection(), mPath, 0);
        }
    }


    private void initPath(ArrowDirection arrowDirection, Path path, float strokeWidth) {
        switch (arrowDirection) {
            case LEFT:
                if (getCornersRadius() <= 0) {
                    initLeftSquarePath(mRect, path, strokeWidth);
                    break;
                }

                if (strokeWidth > 0 && strokeWidth > getCornersRadius()) {
                    initLeftSquarePath(mRect, path, strokeWidth);
                    break;
                }

                initLeftRoundedPath(mRect, path, strokeWidth);
                break;
            case TOP:
                if (getCornersRadius() <= 0) {
                    initTopSquarePath(mRect, path, strokeWidth);
                    break;
                }

                if (strokeWidth > 0 && strokeWidth > getCornersRadius()) {
                    initTopSquarePath(mRect, path, strokeWidth);
                    break;
                }

                initTopRoundedPath(mRect, path, strokeWidth);

                break;
            case RIGHT:
                if (getCornersRadius() <= 0) {
                    initRightSquarePath(mRect, path, strokeWidth);
                    break;
                }

                if (strokeWidth > 0 && strokeWidth > getCornersRadius()) {
                    initRightSquarePath(mRect, path, strokeWidth);
                    break;
                }

                initRightRoundedPath(mRect, path, strokeWidth);

                break;
            case BOTTOM:
                if (getCornersRadius() <= 0) {
                    initBottomSquarePath(mRect, path, strokeWidth);
                    break;
                }

                if (strokeWidth > 0 && strokeWidth > getCornersRadius()) {
                    initBottomSquarePath(mRect, path, strokeWidth);
                    break;
                }

                initBottomRoundedPath(mRect, path, strokeWidth);
                break;
        }
    }

    private void initLeftRoundedPath(RectF rect, Path path, float strokeWidth) {

        path.moveTo(getArrowWidth() + rect.left + getCornersRadius() + strokeWidth, rect.top + strokeWidth);
        path.lineTo(rect.width() - getCornersRadius() - strokeWidth, rect.top + strokeWidth);
        path.arcTo(new RectF(rect.right - getCornersRadius(), rect.top + strokeWidth, rect.right - strokeWidth,
                getCornersRadius() + rect.top), 270, 90);

        path.lineTo(rect.right - strokeWidth, rect.bottom - getCornersRadius() - strokeWidth);
        path.arcTo(new RectF(rect.right - getCornersRadius(), rect.bottom - getCornersRadius(),
                rect.right - strokeWidth, rect.bottom - strokeWidth), 0, 90);


        path.lineTo(rect.left + getArrowWidth() + getCornersRadius() + strokeWidth, rect.bottom - strokeWidth);


        path.arcTo(new RectF(rect.left + getArrowWidth() + strokeWidth, rect.bottom - getCornersRadius(),
                getCornersRadius() + rect.left + getArrowWidth(), rect.bottom - strokeWidth), 90, 90);

        path.lineTo(rect.left + getArrowWidth() + strokeWidth, getArrowHeight() + getArrowPosition() - (strokeWidth / 2));

        path.lineTo(rect.left + strokeWidth + strokeWidth, getArrowPosition() + getArrowHeight() / 2);


        path.lineTo(rect.left + getArrowWidth() + strokeWidth, getArrowPosition() + (strokeWidth / 2));

        path.lineTo(rect.left + getArrowWidth() + strokeWidth, rect.top + getCornersRadius() + strokeWidth);

        path.arcTo(new RectF(rect.left + getArrowWidth() + strokeWidth, rect.top + strokeWidth, getCornersRadius()
                + rect.left + getArrowWidth(), getCornersRadius() + rect.top), 180, 90);

        path.close();
    }

    private void initLeftSquarePath(RectF rect, Path path, float strokeWidth) {

        path.moveTo(getArrowWidth() + rect.left + strokeWidth, rect.top + strokeWidth);
        path.lineTo(rect.width() - strokeWidth, rect.top + strokeWidth);

        path.lineTo(rect.right - strokeWidth, rect.bottom - strokeWidth);

        path.lineTo(rect.left + getArrowWidth() + strokeWidth, rect.bottom - strokeWidth);


        path.lineTo(rect.left + getArrowWidth() + strokeWidth, getArrowHeight() + getArrowPosition() - (strokeWidth / 2));
        path.lineTo(rect.left + strokeWidth + strokeWidth, getArrowPosition() + getArrowHeight() / 2);
        path.lineTo(rect.left + getArrowWidth() + strokeWidth, getArrowPosition() + (strokeWidth / 2));

        path.lineTo(rect.left + getArrowWidth() + strokeWidth, rect.top + strokeWidth);


        path.close();
    }


    private void initTopRoundedPath(RectF rect, Path path, float strokeWidth) {
        path.moveTo(rect.left + Math.min(getArrowPosition(), getCornersRadius()) + strokeWidth, rect.top + getArrowHeight() + strokeWidth);
        path.lineTo(rect.left + getArrowPosition() + (strokeWidth / 2), rect.top + getArrowHeight() + strokeWidth);
        path.lineTo(rect.left + getArrowWidth() / 2 + getArrowPosition(), rect.top + strokeWidth + strokeWidth);
        path.lineTo(rect.left + getArrowWidth() + getArrowPosition() - (strokeWidth / 2), rect.top + getArrowHeight() + strokeWidth);
        path.lineTo(rect.right - getCornersRadius() - strokeWidth, rect.top + getArrowHeight() + strokeWidth);

        path.arcTo(new RectF(rect.right - getCornersRadius(),
                rect.top + getArrowHeight() + strokeWidth, rect.right - strokeWidth, getCornersRadius() + rect.top + getArrowHeight()), 270, 90);
        path.lineTo(rect.right - strokeWidth, rect.bottom - getCornersRadius() - strokeWidth);

        path.arcTo(new RectF(rect.right - getCornersRadius(), rect.bottom - getCornersRadius(),
                rect.right - strokeWidth, rect.bottom - strokeWidth), 0, 90);
        path.lineTo(rect.left + getCornersRadius() + strokeWidth, rect.bottom - strokeWidth);

        path.arcTo(new RectF(rect.left + strokeWidth, rect.bottom - getCornersRadius(),
                getCornersRadius() + rect.left, rect.bottom - strokeWidth), 90, 90);

        path.lineTo(rect.left + strokeWidth, rect.top + getArrowHeight() + getCornersRadius() + strokeWidth);

        path.arcTo(new RectF(rect.left + strokeWidth, rect.top + getArrowHeight() + strokeWidth, getCornersRadius()
                + rect.left, getCornersRadius() + rect.top + getArrowHeight()), 180, 90);

        path.close();
    }

    private void initTopSquarePath(RectF rect, Path path, float strokeWidth) {
        path.moveTo(rect.left + getArrowPosition() + strokeWidth, rect.top + getArrowHeight() + strokeWidth);

        path.lineTo(rect.left + getArrowPosition() + (strokeWidth / 2), rect.top + getArrowHeight() + strokeWidth);
        path.lineTo(rect.left + getArrowWidth() / 2 + getArrowPosition(), rect.top + strokeWidth + strokeWidth);
        path.lineTo(rect.left + getArrowWidth() + getArrowPosition() - (strokeWidth / 2), rect.top + getArrowHeight() + strokeWidth);
        path.lineTo(rect.right - strokeWidth, rect.top + getArrowHeight() + strokeWidth);

        path.lineTo(rect.right - strokeWidth, rect.bottom - strokeWidth);

        path.lineTo(rect.left + strokeWidth, rect.bottom - strokeWidth);


        path.lineTo(rect.left + strokeWidth, rect.top + getArrowHeight() + strokeWidth);

        path.lineTo(rect.left + getArrowPosition() + strokeWidth, rect.top + getArrowHeight() + strokeWidth);


        path.close();
    }


    private void initRightRoundedPath(RectF rect, Path path, float strokeWidth) {

        path.moveTo(rect.left + getCornersRadius() + strokeWidth, rect.top + strokeWidth);
        path.lineTo(rect.width() - getCornersRadius() - getArrowWidth() - strokeWidth, rect.top + strokeWidth);
        path.arcTo(new RectF(rect.right - getCornersRadius() - getArrowWidth(),
                rect.top + strokeWidth, rect.right - getArrowWidth() - strokeWidth, getCornersRadius() + rect.top), 270, 90);

        path.lineTo(rect.right - getArrowWidth() - strokeWidth, getArrowPosition() + (strokeWidth / 2));
        path.lineTo(rect.right - strokeWidth - strokeWidth, getArrowPosition() + getArrowHeight() / 2);
        path.lineTo(rect.right - getArrowWidth() - strokeWidth, getArrowPosition() + getArrowHeight() - (strokeWidth / 2));
        path.lineTo(rect.right - getArrowWidth() - strokeWidth, rect.bottom - getCornersRadius() - strokeWidth);

        path.arcTo(new RectF(rect.right - getCornersRadius() - getArrowWidth(), rect.bottom - getCornersRadius(),
                rect.right - getArrowWidth() - strokeWidth, rect.bottom - strokeWidth), 0, 90);
        path.lineTo(rect.left + getArrowWidth() + strokeWidth, rect.bottom - strokeWidth);

        path.arcTo(new RectF(rect.left + strokeWidth, rect.bottom - getCornersRadius(),
                getCornersRadius() + rect.left, rect.bottom - strokeWidth), 90, 90);

        path.arcTo(new RectF(rect.left + strokeWidth, rect.top + strokeWidth, getCornersRadius()
                + rect.left, getCornersRadius() + rect.top), 180, 90);
        path.close();
    }

    private void initRightSquarePath(RectF rect, Path path, float strokeWidth) {

        path.moveTo(rect.left + strokeWidth, rect.top + strokeWidth);
        path.lineTo(rect.width() - getArrowWidth() - strokeWidth, rect.top + strokeWidth);

        path.lineTo(rect.right - getArrowWidth() - strokeWidth, getArrowPosition() + (strokeWidth / 2));
        path.lineTo(rect.right - strokeWidth - strokeWidth, getArrowPosition() + getArrowHeight() / 2);
        path.lineTo(rect.right - getArrowWidth() - strokeWidth, getArrowPosition() + getArrowHeight() - (strokeWidth / 2));

        path.lineTo(rect.right - getArrowWidth() - strokeWidth, rect.bottom - strokeWidth);

        path.lineTo(rect.left + strokeWidth, rect.bottom - strokeWidth);
        path.lineTo(rect.left + strokeWidth, rect.top + strokeWidth);

        path.close();
    }


    private void initBottomRoundedPath(RectF rect, Path path, float strokeWidth) {

        path.moveTo(rect.left + getCornersRadius() + strokeWidth, rect.top + strokeWidth);
        path.lineTo(rect.width() - getCornersRadius() - strokeWidth, rect.top + strokeWidth);
        path.arcTo(new RectF(rect.right - getCornersRadius(),
                rect.top + strokeWidth, rect.right - strokeWidth, getCornersRadius() + rect.top), 270, 90);

        path.lineTo(rect.right - strokeWidth, rect.bottom - getArrowHeight() - getCornersRadius() - strokeWidth);
        path.arcTo(new RectF(rect.right - getCornersRadius(), rect.bottom - getCornersRadius() - getArrowHeight(),
                rect.right - strokeWidth, rect.bottom - getArrowHeight() - strokeWidth), 0, 90);

        path.lineTo(rect.left + getArrowWidth() + getArrowPosition() - (strokeWidth / 2), rect.bottom - getArrowHeight() - strokeWidth);
        path.lineTo(rect.left + getArrowPosition() + getArrowWidth() / 2, rect.bottom - strokeWidth - strokeWidth);
        path.lineTo(rect.left + getArrowPosition() + (strokeWidth / 2), rect.bottom - getArrowHeight() - strokeWidth);
        path.lineTo(rect.left + Math.min(getCornersRadius(), getArrowPosition()) + strokeWidth, rect.bottom - getArrowHeight() - strokeWidth);

        path.arcTo(new RectF(rect.left + strokeWidth, rect.bottom - getCornersRadius() - getArrowHeight(),
                getCornersRadius() + rect.left, rect.bottom - getArrowHeight() - strokeWidth), 90, 90);
        path.lineTo(rect.left + strokeWidth, rect.top + getCornersRadius() + strokeWidth);
        path.arcTo(new RectF(rect.left + strokeWidth, rect.top + strokeWidth, getCornersRadius()
                + rect.left, getCornersRadius() + rect.top), 180, 90);
        path.close();
    }

    private void initBottomSquarePath(RectF rect, Path path, float strokeWidth) {

        path.moveTo(rect.left + strokeWidth, rect.top + strokeWidth);
        path.lineTo(rect.right - strokeWidth, rect.top + strokeWidth);
        path.lineTo(rect.right - strokeWidth, rect.bottom - getArrowHeight() - strokeWidth);


        path.lineTo(rect.left + getArrowWidth() + getArrowPosition() - (strokeWidth / 2), rect.bottom - getArrowHeight() - strokeWidth);
        path.lineTo(rect.left + getArrowPosition() + getArrowWidth() / 2, rect.bottom - strokeWidth - strokeWidth);
        path.lineTo(rect.left + getArrowPosition() + (strokeWidth / 2), rect.bottom - getArrowHeight() - strokeWidth);
        path.lineTo(rect.left + getArrowPosition() + strokeWidth, rect.bottom - getArrowHeight() - strokeWidth);


        path.lineTo(rect.left + strokeWidth, rect.bottom - getArrowHeight() - strokeWidth);
        path.lineTo(rect.left + strokeWidth, rect.top + strokeWidth);
        path.close();
    }

}

File: app-kotlin/src/main/java/example/shuyu/recycler/kotlin/chat/detail/view/ImageBubbleLayout.kt
package example.shuyu.recycler.kotlin.chat.detail.view

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import android.graphics.Region
import android.util.AttributeSet
import android.view.View

import com.daasuu.bl.ArrowDirection
import com.daasuu.bl.BubbleLayout

/**
 * 转为图片转换的布局
 */
open class ImageBubbleLayout @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : BubbleLayout(context, attrs, defStyleAttr) {

    private var mRect: RectF? = null
    private val mPath = Path()
    private val mPaint = Paint(Paint.ANTI_ALIAS_FLAG)
    private var mStrokePath: Path? = null
    private var mStrokePaint: Paint? = null

    init {
        mPaint.color = bubbleColor
    }


    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        var paddingLeft = paddingLeft
        var paddingRight = paddingRight
        var paddingTop = paddingTop
        var paddingBottom = paddingBottom
        when (arrowDirection) {
            ArrowDirection.LEFT -> paddingLeft -= arrowWidth.toInt()
            ArrowDirection.RIGHT -> paddingRight -= arrowWidth.toInt()
            ArrowDirection.TOP -> paddingTop -= arrowHeight.toInt()
            ArrowDirection.BOTTOM -> paddingBottom -= arrowHeight.toInt()
        }
        setPadding(paddingLeft, paddingTop, paddingRight, paddingBottom)
        initDrawable(0, width, 0, height)
    }


    override fun drawChild(canvas: Canvas, child: View, drawingTime: Long): Boolean {
        drawPath(canvas)
        return super.drawChild(canvas, child, drawingTime)
    }

    private fun drawPath(canvas: Canvas) {
        if (strokeWidth > 0) {
            canvas.drawPath(mStrokePath!!, mStrokePaint!!)
        }
        canvas.clipPath(mPath, Region.Op.INTERSECT)
    }


    private fun initDrawable(left: Int, right: Int, top: Int, bottom: Int) {
        if (right < left || bottom < top) return
        mRect = RectF(left.toFloat(), top.toFloat(), right.toFloat(), bottom.toFloat())
        if (strokeWidth > 0) {
            mStrokePaint = Paint(Paint.ANTI_ALIAS_FLAG)
            mStrokePaint!!.color = strokeColor
            mStrokePath = Path()

            initPath(arrowDirection, mPath, strokeWidth)
            initPath(arrowDirection, mStrokePath!!, 0f)
        } else {
            initPath(arrowDirection, mPath, 0f)
        }
    }


    private fun initPath(arrowDirection: ArrowDirection, path: Path, strokeWidth: Float) {
        when (arrowDirection) {
            ArrowDirection.LEFT -> {
                if (cornersRadius <= 0) {
                    initLeftSquarePath(mRect, path, strokeWidth)
                }

                if (strokeWidth > 0 && strokeWidth > cornersRadius) {
                    initLeftSquarePath(mRect, path, strokeWidth)
                }

                initLeftRoundedPath(mRect, path, strokeWidth)
            }
            ArrowDirection.TOP -> {
                if (cornersRadius <= 0) {
                    initTopSquarePath(mRect, path, strokeWidth)
                }

                if (strokeWidth > 0 && strokeWidth > cornersRadius) {
                    initTopSquarePath(mRect, path, strokeWidth)
                }

                initTopRoundedPath(mRect, path, strokeWidth)
            }
            ArrowDirection.RIGHT -> {
                if (cornersRadius <= 0) {
                    initRightSquarePath(mRect, path, strokeWidth)
                }

                if (strokeWidth > 0 && strokeWidth > cornersRadius) {
                    initRightSquarePath(mRect, path, strokeWidth)
                }

                initRightRoundedPath(mRect, path, strokeWidth)
            }
            ArrowDirection.BOTTOM -> {
                if (cornersRadius <= 0) {
                    initBottomSquarePath(mRect, path, strokeWidth)
                }

                if (strokeWidth > 0 && strokeWidth > cornersRadius) {
                    initBottomSquarePath(mRect, path, strokeWidth)
                }

                initBottomRoundedPath(mRect, path, strokeWidth)
            }
        }
    }

    private fun initLeftRoundedPath(rect: RectF?, path: Path, strokeWidth: Float) {

        path.moveTo(arrowWidth + rect!!.left + cornersRadius + strokeWidth, rect.top + strokeWidth)
        path.lineTo(rect.width() - cornersRadius - strokeWidth, rect.top + strokeWidth)
        path.arcTo(RectF(rect.right - cornersRadius, rect.top + strokeWidth, rect.right - strokeWidth,
                cornersRadius + rect.top), 270f, 90f)

        path.lineTo(rect.right - strokeWidth, rect.bottom - cornersRadius - strokeWidth)
        path.arcTo(RectF(rect.right - cornersRadius, rect.bottom - cornersRadius,
                rect.right - strokeWidth, rect.bottom - strokeWidth), 0f, 90f)


        path.lineTo(rect.left + arrowWidth + cornersRadius + strokeWidth, rect.bottom - strokeWidth)


        path.arcTo(RectF(rect.left + arrowWidth + strokeWidth, rect.bottom - cornersRadius,
                cornersRadius + rect.left + arrowWidth, rect.bottom - strokeWidth), 90f, 90f)

        path.lineTo(rect.left + arrowWidth + strokeWidth, arrowHeight + arrowPosition - strokeWidth / 2)

        path.lineTo(rect.left + strokeWidth + strokeWidth, arrowPosition + arrowHeight / 2)


        path.lineTo(rect.left + arrowWidth + strokeWidth, arrowPosition + strokeWidth / 2)

        path.lineTo(rect.left + arrowWidth + strokeWidth, rect.top + cornersRadius + strokeWidth)

        path.arcTo(RectF(rect.left + arrowWidth + strokeWidth, rect.top + strokeWidth, cornersRadius
                + rect.left + arrowWidth, cornersRadius + rect.top), 180f, 90f)

        path.close()
    }

    private fun initLeftSquarePath(rect: RectF?, path: Path, strokeWidth: Float) {

        path.moveTo(arrowWidth + rect!!.left + strokeWidth, rect.top + strokeWidth)
        path.lineTo(rect.width() - strokeWidth, rect.top + strokeWidth)

        path.lineTo(rect.right - strokeWidth, rect.bottom - strokeWidth)

        path.lineTo(rect.left + arrowWidth + strokeWidth, rect.bottom - strokeWidth)


        path.lineTo(rect.left + arrowWidth + strokeWidth, arrowHeight + arrowPosition - strokeWidth / 2)
        path.lineTo(rect.left + strokeWidth + strokeWidth, arrowPosition + arrowHeight / 2)
        path.lineTo(rect.left + arrowWidth + strokeWidth, arrowPosition + strokeWidth / 2)

        path.lineTo(rect.left + arrowWidth + strokeWidth, rect.top + strokeWidth)


        path.close()
    }


    private fun initTopRoundedPath(rect: RectF?, path: Path, strokeWidth: Float) {
        path.moveTo(rect!!.left + Math.min(arrowPosition, cornersRadius) + strokeWidth, rect.top + arrowHeight + strokeWidth)
        path.lineTo(rect.left + arrowPosition + strokeWidth / 2, rect.top + arrowHeight + strokeWidth)
        path.lineTo(rect.left + arrowWidth / 2 + arrowPosition, rect.top + strokeWidth + strokeWidth)
        path.lineTo(rect.left + arrowWidth + arrowPosition - strokeWidth / 2, rect.top + arrowHeight + strokeWidth)
        path.lineTo(rect.right - cornersRadius - strokeWidth, rect.top + arrowHeight + strokeWidth)

        path.arcTo(RectF(rect.right - cornersRadius,
                rect.top + arrowHeight + strokeWidth, rect.right - strokeWidth, cornersRadius + rect.top + arrowHeight), 270f, 90f)
        path.lineTo(rect.right - strokeWidth, rect.bottom - cornersRadius - strokeWidth)

        path.arcTo(RectF(rect.right - cornersRadius, rect.bottom - cornersRadius,
                rect.right - strokeWidth, rect.bottom - strokeWidth), 0f, 90f)
        path.lineTo(rect.left + cornersRadius + strokeWidth, rect.bottom - strokeWidth)

        path.arcTo(RectF(rect.left + strokeWidth, rect.bottom - cornersRadius,
                cornersRadius + rect.left, rect.bottom - strokeWidth), 90f, 90f)

        path.lineTo(rect.left + strokeWidth, rect.top + arrowHeight + cornersRadius + strokeWidth)

        path.arcTo(RectF(rect.left + strokeWidth, rect.top + arrowHeight + strokeWidth, cornersRadius + rect.left, cornersRadius + rect.top + arrowHeight), 180f, 90f)

        path.close()
    }

    private fun initTopSquarePath(rect: RectF?, path: Path, strokeWidth: Float) {
        path.moveTo(rect!!.left + arrowPosition + strokeWidth, rect.top + arrowHeight + strokeWidth)

        path.lineTo(rect.left + arrowPosition + strokeWidth / 2, rect.top + arrowHeight + strokeWidth)
        path.lineTo(rect.left + arrowWidth / 2 + arrowPosition, rect.top + strokeWidth + strokeWidth)
        path.lineTo(rect.left + arrowWidth + arrowPosition - strokeWidth / 2, rect.top + arrowHeight + strokeWidth)
        path.lineTo(rect.right - strokeWidth, rect.top + arrowHeight + strokeWidth)

        path.lineTo(rect.right - strokeWidth, rect.bottom - strokeWidth)

        path.lineTo(rect.left + strokeWidth, rect.bottom - strokeWidth)


        path.lineTo(rect.left + strokeWidth, rect.top + arrowHeight + strokeWidth)

        path.lineTo(rect.left + arrowPosition + strokeWidth, rect.top + arrowHeight + strokeWidth)


        path.close()
    }


    private fun initRightRoundedPath(rect: RectF?, path: Path, strokeWidth: Float) {

        path.moveTo(rect!!.left + cornersRadius + strokeWidth, rect.top + strokeWidth)
        path.lineTo(rect.width() - cornersRadius - arrowWidth - strokeWidth, rect.top + strokeWidth)
        path.arcTo(RectF(rect.right - cornersRadius - arrowWidth,
                rect.top + strokeWidth, rect.right - arrowWidth - strokeWidth, cornersRadius + rect.top), 270f, 90f)

        path.lineTo(rect.right - arrowWidth - strokeWidth, arrowPosition + strokeWidth / 2)
        path.lineTo(rect.right - strokeWidth - strokeWidth, arrowPosition + arrowHeight / 2)
        path.lineTo(rect.right - arrowWidth - strokeWidth, arrowPosition + arrowHeight - strokeWidth / 2)
        path.lineTo(rect.right - arrowWidth - strokeWidth, rect.bottom - cornersRadius - strokeWidth)

        path.arcTo(RectF(rect.right - cornersRadius - arrowWidth, rect.bottom - cornersRadius,
                rect.right - arrowWidth - strokeWidth, rect.bottom - strokeWidth), 0f, 90f)
        path.lineTo(rect.left + arrowWidth + strokeWidth, rect.bottom - strokeWidth)

        path.arcTo(RectF(rect.left + strokeWidth, rect.bottom - cornersRadius,
                cornersRadius + rect.left, rect.bottom - strokeWidth), 90f, 90f)

        path.arcTo(RectF(rect.left + strokeWidth, rect.top + strokeWidth, cornersRadius + rect.left, cornersRadius + rect.top), 180f, 90f)
        path.close()
    }

    private fun initRightSquarePath(rect: RectF?, path: Path, strokeWidth: Float) {

        path.moveTo(rect!!.left + strokeWidth, rect.top + strokeWidth)
        path.lineTo(rect.width() - arrowWidth - strokeWidth, rect.top + strokeWidth)

        path.lineTo(rect.right - arrowWidth - strokeWidth, arrowPosition + strokeWidth / 2)
        path.lineTo(rect.right - strokeWidth - strokeWidth, arrowPosition + arrowHeight / 2)
        path.lineTo(rect.right - arrowWidth - strokeWidth, arrowPosition + arrowHeight - strokeWidth / 2)

        path.lineTo(rect.right - arrowWidth - strokeWidth, rect.bottom - strokeWidth)

        path.lineTo(rect.left + strokeWidth, rect.bottom - strokeWidth)
        path.lineTo(rect.left + strokeWidth, rect.top + strokeWidth)

        path.close()
    }


    private fun initBottomRoundedPath(rect: RectF?, path: Path, strokeWidth: Float) {

        path.moveTo(rect!!.left + cornersRadius + strokeWidth, rect.top + strokeWidth)
        path.lineTo(rect.width() - cornersRadius - strokeWidth, rect.top + strokeWidth)
        path.arcTo(RectF(rect.right - cornersRadius,
                rect.top + strokeWidth, rect.right - strokeWidth, cornersRadius + rect.top), 270f, 90f)

        path.lineTo(rect.right - strokeWidth, rect.bottom - arrowHeight - cornersRadius - strokeWidth)
        path.arcTo(RectF(rect.right - cornersRadius, rect.bottom - cornersRadius - arrowHeight,
                rect.right - strokeWidth, rect.bottom - arrowHeight - strokeWidth), 0f, 90f)

        path.lineTo(rect.left + arrowWidth + arrowPosition - strokeWidth / 2, rect.bottom - arrowHeight - strokeWidth)
        path.lineTo(rect.left + arrowPosition + arrowWidth / 2, rect.bottom - strokeWidth - strokeWidth)
        path.lineTo(rect.left + arrowPosition + strokeWidth / 2, rect.bottom - arrowHeight - strokeWidth)
        path.lineTo(rect.left + Math.min(cornersRadius, arrowPosition) + strokeWidth, rect.bottom - arrowHeight - strokeWidth)

        path.arcTo(RectF(rect.left + strokeWidth, rect.bottom - cornersRadius - arrowHeight,
                cornersRadius + rect.left, rect.bottom - arrowHeight - strokeWidth), 90f, 90f)
        path.lineTo(rect.left + strokeWidth, rect.top + cornersRadius + strokeWidth)
        path.arcTo(RectF(rect.left + strokeWidth, rect.top + strokeWidth, cornersRadius + rect.left, cornersRadius + rect.top), 180f, 90f)
        path.close()
    }

    private fun initBottomSquarePath(rect: RectF?, path: Path, strokeWidth: Float) {

        path.moveTo(rect!!.left + strokeWidth, rect.top + strokeWidth)
        path.lineTo(rect.right - strokeWidth, rect.top + strokeWidth)
        path.lineTo(rect.right - strokeWidth, rect.bottom - arrowHeight - strokeWidth)


        path.lineTo(rect.left + arrowWidth + arrowPosition - strokeWidth / 2, rect.bottom - arrowHeight - strokeWidth)
        path.lineTo(rect.left + arrowPosition + arrowWidth / 2, rect.bottom - strokeWidth - strokeWidth)
        path.lineTo(rect.left + arrowPosition + strokeWidth / 2, rect.bottom - arrowHeight - strokeWidth)
        path.lineTo(rect.left + arrowPosition + strokeWidth, rect.bottom - arrowHeight - strokeWidth)


        path.lineTo(rect.left + strokeWidth, rect.bottom - arrowHeight - strokeWidth)
        path.lineTo(rect.left + strokeWidth, rect.top + strokeWidth)
        path.close()
    }

}
